
\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}
%customized
\linespread{1.5}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8x]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% color can be used to apply background shading to table cells only
\usepackage[table]{xcolor}

% array package and thick rules for tables
\usepackage{array}

% create "+" rule type for thick vertical lines
\newcolumntype{+}{!{\vrule width 2pt}}

% create \thickcline for thick horizontal lines of variable length
\newlength\savedwidth
\newcommand\thickcline[1]{%
  \noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
  \cline{#1}%
  \noalign{\vskip\arrayrulewidth}%
  \noalign{\global\arrayrulewidth\savedwidth}%
}

% \thickhline command for thick horizontal lines that span the table
\newcommand\thickhline{\noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
\hline
\noalign{\global\arrayrulewidth\savedwidth}}

% Remove comment for double spacing
%\usepackage{setspace} 
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}
\renewcommand{\figurename}{Fig}

% Use the PLoS provided BiBTeX style
\bibliographystyle{plos2015}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% Leave date blank
\date{}

% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\setlength{\headheight}{27.023pt}
\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\sf PLOS}

%units
\usepackage{siunitx}
\DeclareSIUnit\Molar{\textsc{m}}
%\sisetup{scientific-notation = true}
%% Include all macros below

\newcommand{\lorem}{{\bf LOREM}}
\newcommand{\ipsum}{{\bf IPSUM}}

%% END MACROS SECTION


\begin{document}
\vspace*{0.2in}

% Title must be 250 characters or less.
\begin{flushleft}
{\Large
\textbf\newline{Simulating detailed Ca\textsuperscript{2+}-calmodulin-CaMKII network in a new way: biophysical attributes that affect CaMKII activation pattern} % Please use "sentence case" for title and headings (capitalize only the first word in a title (or heading), the first word in a subtitle (or subheading), and any proper nouns).
}
\newline
% Insert author names, affiliations and corresponding author email (do not include titles, positions, or degrees).
\\
Ximing Li\textsuperscript{1},
William Holmes\textsuperscript{1}
\\
\bigskip
\textbf{1} Department of Biological Sciences, Neuroscience Program, Ohio University, Athens, Ohio, United States of America
\\
\bigskip

% Insert additional author notes using the symbols described below. Insert symbol callouts after author names as necessary.
% 
% Remove or comment out the author notes below if they aren't used.
%
% Current address notes
%å\textcurrency Current Address: Dept/Program/Center, Institution Name, City, State, Country % change symbol to "\textcurrency a" if more than one current address note
% \textcurrency b Insert second current address 
% \textcurrency c Insert third current address

% Deceased author note
%\dag Deceased

% % Group/Consortium Author Note
% \textpilcrow Membership list can be found in the Acknowledgments section.

% % Use the asterisk to denote corresponding authorship and provide email address in note below.
% * correspondingauthor@institute.edu
\end{flushleft}
% Please keep the abstract below 300 words
\section*{Abstract}
Calcium/calmodulin-dependent protein kinase II (CaMKII) holoenzymes play a critical role in decoding Ca\textsuperscript{2+} signals in neurons and understanding how this occurs has been the focus of many modeling studies. However, CaMKII is notoriously difficult to model and simulate in detail because of its multi-subunit nature, which causes a combinatorial explosion of the number of species that must be modeled. To study the Ca\textsuperscript{2+}-calmodulin-CaMKII reaction network with detailed kinetics while including the effect of diffusion, we have customized an existing stochastic particle-based simulator Smoldyn to manage the problem of combinatorial explosion. With this new method, spatial and temporal aspects of the signaling network can be studied without compromising biochemical details. We used this new method to examine how calmodulin molecules, both partially loaded and fully loaded with Ca\textsuperscript{2+}, choose pathways to interact with and activate CaMKII under various Ca\textsuperscript{2+} input conditions. We found that the frequency dependence of the CaMKII phosphorylation is intrinsic to the network kinetics and the activation pattern can be modulated by the relative amount of Ca\textsuperscript{2+} to calmodulin and by the rate of Ca\textsuperscript{2+} diffusion. Depending on whether Ca\textsuperscript{2+} influx is saturating or not, calmodulin molecules could choose different routes within the network to activate CaMKII subunits, resulting in different frequency dependence patterns. In addition, the size of the holoenzyme shows a subtle effect on CaMKII activation. The more extended the subunits are organized, the easier for CaM molecules to access and activate the subunits. The findings suggest that particular intracellular environmental factors such as crowding and calmodulin availability can play an important role in decoding Ca\textsuperscript{2+} signals and can give rise to distinct CaMKII activation patterns in dendritic spines, Ca\textsuperscript{2+} channel nanodomains and cytoplasm.
% Please keep the Author Summary between 150 and 200 words
% Use first person. PLOS ONE authors please skip this step. 
% Author Summary not valid for PLOS ONE submissions.   
\section*{Author summary}

Ca\textsuperscript{2+} signals are commonly used by cells for various types of activities. In neurons, Ca\textsuperscript{2+} can regulate gene expression and dendritic spine enlargement, strengthen synaptic connectivity, and promote neural growth or even death. These strikingly contrasting cell processes are possible because Ca\textsuperscript{2+} signals of varying strengths and temporal patterns are interpreted differently by intracellular signaling pathways. In our study, we focus on one particular pathway that involves calmodulin and Ca\textsuperscript{2+}/calmodulin-dependent protein kinase II (CaMKII). We have developed a new computational method that avoids the problem of combinatorial explosion inherent with modeling this pathway in biophysical detail, and this allows us to model this pathway more realistically than conventional methods. Experiments have shown that CaMKII activity is sensitive to Ca\textsuperscript{2+} signal frequency and our models demonstrate how this frequency dependence depends on Ca\textsuperscript{2+} input conditions, calmodulin availability and the Ca\textsuperscript{2+} diffusion rate.

\linenumbers
% Use "Eq" instead of "Equation" for equation citations.
% Camkii binding with F-actin is mostly reported in spines studies, but also seen in cytoplasm as Li Boxing 2016
%Bhalla's paper CaMKII model with diffusion, but simplified CaM states;
% setup with spine geometry; camca4 only; no diffusion effect; stochastic? hybrid? deterministic
\section*{Introduction}
Calcium/calmodulin-dependent protein kinase II (CaMKII) is an important enzyme widely distributed in the central nervous system and other tissues including cardiac muscles~\cite{Wayman:2008gla,Lisman:2002ki,Erickson:2014fs}. CaMKII is involved in intracellular Ca\textsuperscript{2+} signaling as an effector of the Ca\textsuperscript{2+} sensor protein calmodulin (CaM)~\cite{Herring:2016bh,Hell:2014bd,Coultrap:2012ip}. In neurons, CaMKII molecules are essential in long-term potentiation (LTP), which is our best model for the molecular basis of learning and memory~\cite{Bliss:1973jg}. In spines, activated CaMKII molecules interact with postsynaptic density proteins, facilitating actins to reorganize, leading to spine enlargement and upregulation of $alpha$-amino-3-hydroxy-5-methyl-4-isoxazolepropionic acid (AMPA) receptor numbers~\cite{Herring:2016bh}. CaMKII molecules can also phosphorylate AMPA receptors to regulate channel conductance~\cite{Herring:2016bh,Coultrap:2012ip} or form complexes with NMDA receptors~\cite{Hell:2014bd}. In addition to synaptic functions, recent studies~\cite{Ma:2014dr,Li:2016cq} suggest that CaMKII in the soma can act as carriers to shuttle Ca\textsuperscript{2+}-CaM into the nucleus. Ca\textsuperscript{2+}-CaM molecules are then unloaded in the nucleus, participate in the calcium/calmodulin-dependent protein kinase IV (CaMKIV) cascade to activate nuclear transcription factors. Therefore CaMKII molecules play a key role in excitation-transcription coupling. 

The CaMKII molecule is a holoenzyme that consists of 12 subunits grouped in 2 rings. Each subunit contains an association domain allowing the formation of multimers, a regulatory domain with CaM binding sites and phosphorylation sites, and a catalytic domain to act as a kinase. Activities such as CaM binding and auto-phosphorylation result in various subunits states. For example, in the presence of Ca\textsuperscript{2+}-CaM complexes, each subunit can bind a CaM molecule and expose its catalytic domain to phosphorylate the direct neighbor subunit at its Thr286/Thr287 site. Once phosphorylated, the CaM unbinding rate decreases dramatcially, leading to a prolonged activation of the subunit; we say, the CaM molecule is "trapped" by the CaMKII subunit~\cite{Meyer:1992dp}. Even when the CaM molecule unbinds, the CaMKII subunit stays activated, entering an autonomous state. Finally, Thr305/Thr306 can also be phosphorylated. There theronine sites overlap with the CaM binding site and its phosphorylation blocks the binding of CaM molecules. In this case, the subunit becomes "capped". 

Researchers have long been interested in modeling and simulating CaMKII~\cite{Zhabotinsky:2000fp,Holmes:2000uk,Kubota:2001ul,Dupont:2003vq,Bhalla:2004cu,Lucic:2008gt,Zeng:2010bq,Pepke:2010ju,Michalski:2012ds}. However, the structural complexity and multi-state nature of CaMKII present a technical challenge. A major problem is combinatorial explosion~\cite{Stefan:2014gl}. One CaMKII subunit has only a countable number of states to consider, but with 6 or 12 subunits on a holoenzyme, the number of combinations of states for a holoenzyme becomes extremely large. This is a common problem in systems biology since large proteins usually consist of multiple subunits. Software that adopts a rule-based approach such as BioNetGen~\cite{Hlavacek:2006iq} can be used to expand the network based on a set of given reaction rules. But for CaMKII holoenzyme, using BioNetGen to generate the network is still computationally intensive. After expansion with BioNetGen, a 6-subunit holoenzyme with four states for each subunit contains 700 unique species and 12192 unique reactions. The size of the network grows substantially when more detailed kinetics are involved. For example, each CaM has 4 Ca\textsuperscript{2+}-binding sites, giving rise to 9 distinct binding states; each state has its own distinct kinetics to interact with a CaMKII subunit. Then each CaMKII subunit exhibits a distinct phosphorylation rate depending on the state of CaM being bound. For simplicity, consider a 6-subunit CaMKII holoenzyme. Each subunit would potentially have 20 states. Then for the holoenzyme, the total number of unique species reaches over $\frac{6^{20}}{6}=609.36\times10^{12}$. It would be extremely time-consuming and not practical for BioNetGen to generate this network. To overcome this problem, most previous modeling studies of CaMKII simplified the Ca\textsuperscript{2+}-CaM-CaMKII reaction network by either modeling CaMKII as monomers~\cite{Bhalla:2004cu,Pepke:2010ju} or allowing only CaM fully-loaded with Ca\textsuperscript{2+} to interact with CaMKII subunits~\cite{Zhabotinsky:2000fp,Kubota:2001ul,Dupont:2003vq,Michalski:2012ds}.

Another complication is related to the intracellular environment. Conventionally, biochemical reactions are modeled in a deterministic manner using Ordinary Differential Equations (ODEs). The underlying assumptions are that reactions occur in a spatially homogenous environment, reactants are abundant and not subject to stochasticity, and molecules diffuse sufficiently fast. The majority of previous modeling studies belong to this category except for a few that are stochastic or hybrid models ~\cite{Bhalla:2004cu, Zeng:2010bq, Holmes:2000uk}. However, intracellular space is highly heterogeneous and compartmentalized~\cite{LubyPhelps:2000uj,Dix:2008gy}. Reactions are often restricted to a small space. Structures such as the cytoskeleton, scaffold proteins or endoplasmic reticulum often act as diffusion barriers to slow down molecules. In addition, most previous modeling studies focus on the dendritic spine, where NMDA receptor-channels are the main Ca\textsuperscript{2+} providers. 

In the present work, we focus on the soma where voltage-dependent Ca\textsuperscript{2+} channels (CaV) provide the Ca\textsuperscript{2+} influx, which serves for our future study on modeling Ca\textsuperscript{2+}-CaM transportation from surface to the nucleus. Specifically, we examined Ca\textsuperscript{2+}-CaM-CaMKII network activity near L-type Ca\textsuperscript{2+} channels. This is an area less studied previous models, but nonetheless plays a critical role in excitation-transcription coupling~\cite{Ma:2015bg,Li:2016cq}. To study this Ca\textsuperscript{2+}-CaM-CaMKII signaling network and deal with the problem of combinatorial explosion, we modified the published and freely available simulator Smoldyn~\cite{Andrews:2004fs}. Smoldyn is a particle-based stochastic simulator and has been used for simulating reaction and diffusion processes in cells. It works well for relatively simple reaction networks but not for holoenzymes such as CaMKII. We added new data structures to describe the reaction network in a compact way. CaMKII holoenzymes are modeled as a collection of subunits. Each subunit has a set of binding sites. Subunits react independently and diffuse collectively. Reactions are defined between binding sites. The reaction network is stored in a hashtable to allow for lookup during simulations only when reactants collide, without expanding and loading a complete network. We tested and verified these modifications by modeling a Ca\textsuperscript{2+}-CaM interaction network and comparing results with COPASI~\cite{Hoops:2006gy}. Moreover, we added detailed components to the network and examined the frequency dependence of CaMKII activation. We found that slow diffusion of Ca\textsuperscript{2+} can dramatically boost CaMKII activation. Also in the presence of a limited number of CaM molecules, CaM must make an intricate choice between binding Ca\textsuperscript{2+} first or a CaMKII subunit first. Interestingly, a change in the decision pattern leads to an altered frequency dependence of the network. 

% % Results and Discussion can be combined.
\section*{Results}
\subsection*{Testing of the simulator modifications}
We first tested the modifications to Smoldyn used in this paper (see Methods) using a Ca\textsuperscript{2+}-CaM network (Fig~\ref{fig1}A). The 4 Ca\textsuperscript{2+} binding sites on CaM gives rise to a total of 9 different binding states of a CaM molecule. For the reaction volume we use a $\SI{500}{\nm}\times\SI{500}{\nm}\times\SI{500}{\nm}$ cube with all sides being reflective. The size of this volume mimics that of a Ca\textsuperscript{2+} channel nanodomain or a small dendritic spine head. Initially, the cube contains 3000 Ca\textsuperscript{2+} ions and 700 apoCaM molecules, equivalent to \SI{39.867}{\micro\Molar} and 9.302\SI{}{\micro\Molar} respectively. The concentrations are chosen to produce an observable amount of 4-Ca\textsuperscript{2+} bound CaM at the steady state. All molecules are initially uniformly distributed. We tested reactions using two different Ca\textsuperscript{2+} diffusion constants \SI{2.2e-6}{\square\cm\per\s}~\cite{Keller:2008ez} and half of this value \SI{1.1e-6}{\square\cm\per\s}.

We first characterized the reactions in the network to see if results of our stochastic model could reasonably be compared to results with standard ODE methods. A second order chemical reaction in the solvent phase consists of two steps: molecules encountering each other by diffusion followed by the molecules reacting with each other. If the encountering step takes a much longer time to occur than the reacting step, the reaction is diffusion-limited; otherwise, the reaction is activation-limited. Conventionally, an experimentally measured binding kinetic rate, $k_{on}$, can be decomposed into an encounter rate, $k_{enc}$, and an intrinsic activation rate, $k_{a}$, using the following equation~\cite{Bamford198505},
\begin{equation}\frac{1}{k_{on}}=\frac{1}{k_{enc}}+\frac{1}{k_a}\end{equation}
For diffusion-limited reactions, $k_a \gg k_{on}$ and $k_{on} \approx k_{enc}$; for activation-limited reactions, $k_a \approx k_{on}$ as diffusion is sufficiently fast. Given the diffusion coefficients for Ca\textsuperscript{2+} and CaM and the parameter values for Ca\textsuperscript{2+} and CaM reaction kinetics in Table S2, we calculated the $\frac{k_a}{k_{on}}$ ratios and concluded that the reactions in the network belong to the activation-limited regime. This was also true if diffusion constants were reduced 10 fold. Thus this simple model resembles a well-mixed system and a standard ODE model will provide a good test standard. 

We coded a deterministic ODE system in COPASI~\cite{Hoops:2006gy} and compared results with our stochastic model using the same kinetics. We found that the stochastic model and the ODE COPASI model exhibited similar time courses for all CaM binding state species as show in Fig~\ref{fig2}. Slowing down Ca\textsuperscript{2+} diffusion did not significantly alter the time courses for the parameter values used. These results suggest that our modifications of Smoldyn are working properly. 

\begin{figure}[!h]
	\caption{{\bf Ca\textsuperscript{2+}-CaM interaction network.}
	A: Ca\textsuperscript{2+} and CaM interactions can produce 9 different states of CaM. CaM has two lobes, an N lobe and a C lobe, each of which has 2 Ca\textsuperscript{2+} binding sites. Binding sites on the lobes are designated N1, N2, C1 and C2. The N lobe and C lobe bind Ca\textsuperscript{2+} independently of each other, but Ca\textsuperscript{2+} binding within a lobe is cooperative. Thus site N2 cannot bind Ca\textsuperscript{2+} unless N1 has already bound Ca\textsuperscript{2+} and a similar rule applies to sites C2 and C1. In the figure N0C0 represents apoCaM (no Ca\textsuperscript{2+} bound), "N2" implies binding sites N1 and N2 are both bound with Ca\textsuperscript{2+} and "C2" implies similarly that both C1 and C2 have Ca\textsuperscript{2+} bound, "N1" and "C1" imply only the "N1" site and only the "C1" site have Ca\textsuperscript{2+} bound. Arrows represent the direction of binding reactions.
	B: The same reactions between Ca\textsuperscript{2+} and CaM as shown in A are represented using the syntax of the modified simulator. The binding and unbinding rate constants are labeled at the end of reactions. ‘+’ represents a binding reaction; ‘$\sim$’ represents molecules that are bound. ‘\{\}’ on the left-hand side specifies the states of binding sites of reactants or the conditions for a reaction to occur. On the right-hand side, binding sites involved in the reaction are assigned to new values. Notice the ‘==‘ sign is used on the left-hand side, but the ‘=‘ sign is used on the right-hand side. The ‘==‘ represents True or False of equality, whereas the ‘=‘ denotes an assigning operation. 
	}
\label{fig1}
\end{figure}
	 
\begin{figure}[!h]
	\caption{{\bf Test of Smoldyn modifications with a simple Ca\textsuperscript{2+}-CaM reaction network.}
	Plots compare the number of CaM molecules in each of the 9 possible states as computed with the modified stochastic simulator and COPASI. Thick smooth lines are results using COPASI and thin black and red lines are results with the modified simulator for two different values of the calcium diffusion coefficient.
	}
\label{fig2}
\end{figure}

As another verification, we tested the steady state fraction of phosphorylation of CaMKII holoenzymes when they are modeled as multi-subunit complex. As Michalski and Loew~\cite{Michalski:2012ds} pointed out, in a closed system, assuming saturating amount of fully loaded CaM molecules are provided and subunit phosphorylations are allowed only from neighbors that are CaM bound but not those phosphorylated, at the steady state, there will be distinctive phosphorylation levels depending on the number of subunits in the holoenzyme. In particular, for dimers the steady state fraction of phosphorylated subunits is $\frac{1}{2}$; for trimers, $\frac{2}{3}$; for large subunit numebrs, the fraction approaches $1-e^{-1}$. We built a model using a closed system with $\SI{1}{\um}\times\SI{1}{\um}\times\SI{1}{\um}$ geomemtry. 6000 CaMKII subunits and 6020 fully loaded CaM were included in the system. Only the following reactions were allowed: bindings and unbindings between CaM and CaMKII, as well as phosphorylations by neighbor subunits that were bound but not phosphorylated. As expected, we obtained steady state phosphorylation levels that match the Michalski and Loew's formula. Finally, using the same closed system, we tested the complete network and compared with results obtained from a simulation implemented with spatial Gillespie algorithm by Zeng and Holmes~\cite{Zeng:2010bq}. The two simulations showed comparable results.

\subsection*{Analysis of the Ca\textsuperscript{2+}-CaM-CaMKII network}

We set up a prototype model to identify the major network branches that contribute to the phosphorylation of CaMKII subunits. The prototype model comprises Ca\textsuperscript{2+}-CaM-CaMKII interactions as shown in Fig~\ref{fig3}A. In this network diagram molecule species are represented as vertices and reactions are represented as the network edges. CaM molecules and CaMKII holoenzymes with two 6-subunit rings were initially uniformly distributed in a $\SI{1}{\um}\times\SI{1}{\um}\times\SI{2}{\um}$ box. As an initial condition, the box contains 6020 apoCaM molecules (\SI{5}{\micro\Molar}) and 12036 CaMKII subunits (\SI{10}{\micro\Molar}), within which 200 CaM molecules are bound to CaMKII. The initial state is at an equilibrium, which we carefully computed by running 5 simulation trials starting with the same numbers of CaM and CaMKII molecules. For simplicity, we do not consider the resting level intracellular Ca\textsuperscript{2+} contributed significantly to the steady state initial condition. Extracellular Ca\textsuperscript{2+} ions were released as a single source from the top center of box ($\SI{1}{\um}\times$\SI{1}{\square\um}). A previously generated input file having a total of 6 Ca\textsuperscript{2+} bursts delivered at 5 Hz was used to provide Ca\textsuperscript{2+} influx during the simulation (see Methods). We ran the simulation up to \SI{2.5}{\s}, recorded the molecule numbers at every \SI{1}{\ms} and logged all reaction events. Using the events log, we counted the accumulated occurrences of each reaction type at every \SI{10}{\ms}, starting from the initial steady-state until the Ca\textsuperscript{2+} bursts were over. For a particular reaction, the number of occurrences during a time span is considered to be the net number of molecule state changes along the corresponding edge, i.e., the number of unbinding events is subtracted from the number of binding events. A negative number means that unbinding occurs more often than binding within the given time span. 

\begin{figure}[!h]
	\caption{{\bf The complete Ca\textsuperscript{2+}-CaM-CaMKII network.}
	A: Species are represented by vertices. Edges are labeled to indicate binding reactions. For example, "N1\_rxn" means binding of a Ca\textsuperscript{2+} ion on CaM at the N1 site. "KN1\_rxn" means CaM attached to a CaMKII subunit binds a Ca\textsuperscript{2+} ion on its N1 site. "K-NxCy\_rxn" means a CaMKII subunit binds a CaM that has x Ca\textsuperscript{2+} ions bound on the N-lobe and y Ca\textsuperscript{2+} ions bound on the C-lobe (x and y range from 0 to 2). "KpNxCy\_rxn" means a phosphorylated CaMKII subunit binds a CaM with x and y Ca\textsuperscript{2+} ions bound on the N and C lobes respectively. Red edges mark the predominant pathway. B: Layers shaded with blue colors. The back layer(Layer1) represents interactions between Ca\textsuperscript{2+} and CaM. The front layer(Layer2) represents interactions between Ca\textsuperscript{2+} and CaM attached to a CaMKII subunit. Layer3 represents interactions between Ca\textsuperscript{2+} and CaM attached to a phosphorylated CaMKII subunit.
	}
\label{fig3}
\end{figure}

We analyzed the edges in the Ca\textsuperscript{2+}-CaM-CaMKII network (Fig~\ref{fig3}A). The network consists of three layers (Fig~\ref{fig3}B). The back layer (Layer1) describes interactions between Ca\textsuperscript{2+} and the 9 states of CaM (NxCy, x,y=0,1,2). The front layer (Layer2) is for reactions of Ca\textsuperscript{2+} with CaM attached to unphosphorylated CaMKII (KNxCy). The middle layer (Layer3) is for reactions of Ca\textsuperscript{2+} with CaM attached to phosphorylated CaMKII (KpNxCy). Starting in Layer1, there are at most 4 possible ways for each CaM species to change state: binding a Ca\textsuperscript{2+} at a C site, binding Ca\textsuperscript{2+} at an N site, binding to an unphosphorylated CaMKII subunit or binding to a phosphorylated CaMKII subunit. However, binding to a phosphorylated CaMKII subunit requires a bound subunit to be phosphorylated first and then lose the bound CaM. This type of reaction rarely occurs during the early stage of the simulation since phosphorylation is slow and CaM unbinding from a phosphorylated subunit is uncommon. Thus we focus on the first 3 types of reactions.

For each CaM state, we counted the accumulated occurrences of the possible reaction types. Results are shown in Fig~\ref{fig4}. The plot reveals a predominant pathway for CaM and CaMKII state transitions as labeled in Fig~\ref{fig3}A. Starting as apoCaM, a CaM molecule tends to bind a Ca\textsuperscript{2+} ion on the C1 site and then on the C2 site, entering the N0C2 state. CaM in the N0C2 state has a strong preference to enter layer 2 by binding to a CaMKII subunit, after which it quickly binds Ca\textsuperscript{2+} ions to the N1 and N2 sites to become KN2C2. Also evident from Fig~\ref{fig4} is that once CaM has two Ca\textsuperscript{2+} ions on either lobe, the preference is to bind to a CaMKII subunit before binding additional Ca\textsuperscript{2+} ions. Once bound to a CaMKII subunit, additional Ca\textsuperscript{2+} ions bind to CaM more quickly. In this scenario, it is rare for CaM to become fully bound with Ca\textsuperscript{2+} ions prior to binding with a CaMKII subunit. Nevertheless, phosphorylation of CaMKII subunits occurs most often when subunits are bound with fully loaded CaM, but still often with CaM having C sites loaded, KN0C2 and KN1C2, as shown in Fig~\ref{fig5}. This preferred pathway is consistent with a hypothetical major pathway that leads to CaMKII autophosphorylation shown in Pepke et al.~\cite{Pepke:2010ju}. The observation is also consistent with Shifman et al.~\cite{Shifman:2006hw} in showing that partially loaded CaM molecules may be very important in activating CaMKII subunits. We note that the way the predominant pathway is chosen at each vertex is a consequence of reaction affinities as given in Table S2. For example, even though N sites bind Ca\textsuperscript{2+} faster, the C sites have higher affinity making Ca\textsuperscript{2+} binding to C sites preferred. 

\begin{figure}[!h]
	\caption{{\bf Accumulated reaction occurrences over time for Layer1 reactions.}
	For a CaM binding state, 3 types of reactions were shown: binding of Ca\textsuperscript{2+} at an N site, binding of Ca\textsuperscript{2+} at a C site or binding of CaM to an unphosphorylated CaMKII subunit. Green arrows indicate reactions participating in the predominant pathway starting from N0C0.
	}
\label{fig4}
\end{figure}

\begin{figure}[!h]
	\caption{{\bf Accumulated reaction occurrences over time for Layer2 reactions.}
	For a CaM binding state, 3 types of reactions were shown: binding of Ca\textsuperscript{2+} at an N site, binding of Ca\textsuperscript{2+} at a C site or phosphorylation of the CaMKII that has the CaM bound.
	}
\label{fig5}
\end{figure}


To confirm the critical role of NxC2 (x=0,1,2) CaM in activation and phosphorylation of CaMKII, we set up two modified reaction schemes (Fig~\ref{fig6}A). In Scheme1, only CaMKII subunits bound with NxC2 are allowed to become phosphorylated. In Scheme2, phosphorylation is allowed only for subunits bound with N2Cx. Note the phosphorylation rates in the two schemes are equivalent, i.e., $k_{on}$ of reaction KpN2C0\_rxn is the same as that of KpN0C2\_rxn, and the $k_{on}$ of KpN2C1\_rxn equals that of KpN1C2\_rxn. Not surprisingly, the two schemes give rise to different phosphorylation levels as shown in Fig~\ref{fig6}B. Scheme 1 performs slightly worse than the whole network, whereas Scheme 2 produces much lower phosphorylation levels. Therefore, edges of the network are not equally involved in CaMKII phosphorylation. Phosphorylation from KNxC2 is especially important.  

\begin{figure}[!h]
	\caption{{\bf Reaction scheme 1 and 2 adapted from the complete network.}
	A: Reaction scheme 1 and 2 to verify that CaMKII bound with CaM NxC2 are more invovled in phosphorylation than those bound with N2Cx.
	B. Comparison of phosphorylation levels under complete network, scheme 1 and scheme 2. The results of the complete network were summarized from 15 trials of simulations and presented in mean phosphorylation +/- standard deviation.
	} 
\label{fig6}
\end{figure}


\subsection*{CaMKII activation frequency dependence and the detailed Ca\textsuperscript{2+}-CaM-CaMKII network mechanisms}

A classic experiment by DeKonick and Schulman~\cite{DeKoninck:1998wh} showed that \textit{in vitro} CaMKII holoenzyme activation is sensitive to the frequency of Ca\textsuperscript{2+}-CaM pulses. A recent study~\cite{Fujii:2013bg} also showed that \textit{in vivo} glutamate uncaging frequency affects CaMKII activation in dendritic spines. Numerous modeling studies have also reported a dependence of CaMKII activation on the frequency of Ca\textsuperscript{2+} signals~\cite{Pepke:2010ju,Dupont:2003vq,Michalski:2012ds,Kubota:2001ul}. To confirm this CaMKII activation pattern, we tested our 6-subunit CaMKII model using previously generated 10 Hz and 5 Hz Ca\textsuperscript{2+} influx files. The total number of Ca\textsuperscript{2+} ions entering was comparable (40433 ions with 5 Hz vs 40435 ions with 10 Hz). As shown in Fig~\ref{fig7}, we observed much higher phosphorylation levels with 10 Hz than with 5 Hz input.

\begin{figure}[!h]
	\caption{{\bf Comparison of 10 Hz and 5 Hz on CaMKII activation.}
	A: Number of bound CaMKII subunits (including bound and phosphorylated) under 5 Hz (thin line) and 10 Hz (thick line) Ca\textsuperscript{2+} input.
	B: Number of phosphorylated CaMKII subunits (including bound and phosphorylated) under 5 Hz (thin line) and 10 Hz (thick line) Ca\textsuperscript{2+} input.
	}
\label{fig7}
\end{figure}

We examined the reaction occurrences over time and noticed that the 10 Hz input results in more bindings between N0C2 CaM and CaMKII (Fig~\ref{fig8}A,B), and consequently more autophosphorylation (Fig~\ref{fig8}C,D). Therefore the frequency effect is inherent to the binding between Ca\textsuperscript{2+}, CaM and CaMKII. The frequency effect can be reversed by providing saturating amount of Ca\textsuperscript{2+}. As demonstrated in Fig~\ref{fig9}, we increased the Ca\textsuperscript{2+} channel number to allow more Ca\textsuperscript{2+} influx per action potential. The total Ca\textsuperscript{2+} influx was again comparable for 5 Hz and 10 Hz. As Ca\textsuperscript{2+} influx was increased, the network produced more phosphorylated CaMKII subunits for both input frequencies, but the phosphorylation level difference between 5 Hz and 10 Hz diminished. Eventually, the 10 Hz input becomes saturating and the network generates less phosphorylation under 10 Hz than the 5 Hz input (Fig~\ref{fig9}F).
\begin{figure}[!h]
	\caption{{\bf Effects of 10 Hz and 5 Hz on reaction occurrences.} 
	A: Accumulated reaction occurrences for CaM CaMKII binding reactions in the presence of 5 Hz Ca\textsuperscript{2+} input. 
	B: The same as in A but for 10 Hz Ca\textsuperscript{2+} input.
	C: Accumulated reaction occurrences for CaMKII phosphorylation reactions in the presence of 5 Hz Ca\textsuperscript{2+} input. 
	D: The same as in C but for 10 Hz Ca\textsuperscript{2+} input.
	}
\label{fig8}
\end{figure}

\begin{figure}[!h]
	\caption{{\bf Effects of various Ca\textsuperscript{2+} influx on frequency dependence.}
	A-E: Accumulated reaction occurrences for phosphorylation of CaMKII subunits bound with NxC2 in the presence of 5 Hz (thin line) and 10 Hz (thick line) with different levels of Ca\textsuperscript{2+} influx. 
	F: Summary of phosphorylation levels in the presence of varying amount of Ca\textsuperscript{2+} delivered at 5 Hz and 10 Hz. The advantage of 10 Hz on phosphorylation is reversed as Ca\textsuperscript{2+} influx increases.
	}
\label{fig9}
\end{figure}

Equivalently, the saturating Ca\textsuperscript{2+} effects on frequency preference can also be facilitated by limiting the available CaM and slowing Ca\textsuperscript{2+} diffusion (Fig~\ref{fig10}). To demonstrate, we compared simulations using $3.5\times$ Ca\textsuperscript{2+} input with the default amount of CaM (\SI{5}{\micro\Molar}) and limited CaM (\SI{2.5}{\micro\Molar}). Even though the Ca\textsuperscript{2+} input does not change, by limiting CaM we noticed that the 5 Hz input results in more phosphorylation reactions than 10 Hz(Fig~\ref{fig10}A,B). The same effect can be achieved with slowed Ca\textsuperscript{2+} diffusion. We set Ca\textsuperscript{2+} diffusion as \SI{0.0000011}{\square\cm\per\s}, which is half of the default speed and does not change the activation-limited regime of the network. Slow diffusion results in a dramatic increase of CaMKII phosphorylation regardless of Ca\textsuperscript{2+} amount and input frequencies. Also, the frequency preference for 10 Hz is reversed at $3\times$ Ca\textsuperscript{2+} influx condition(Fig~\ref{fig10}C,D). In other words, with slow diffusion, Ca\textsuperscript{2+} influx provided by $3\times$ becomes sufficiently saturating to change the frequency preference. It is also worth noting that by combining limited CaM and slow diffusion, reversing the frequency preference is possible with merely twice amount of Ca\textsuperscript{2+} influx (Fig~\ref{fig10}E,F). 

\begin{figure}[!h]
	\caption{{\bf Effects of CaM availability and Ca\textsuperscript{2+} diffusion on frequency dependence.}
	A: With $3\times$ Ca\textsuperscript{2+} influx, accumulated reaction occurrences for CaMKII phosphorylation bound with NxC2 in the presence of 5 Hz (thin line) and 10 Hz (thick line) Ca\textsuperscript{2+} input.
	B: The same as in A except that the total CaM available in the system is decreased to half. 
	C: The same as in A except that $3.5\times$ Ca\textsuperscript{2+} is provided.
	D: The same as in C except that Ca\textsuperscript{2+} diffusion is slowed  to \SI{0.0000011}{\square\cm\per\s}. 
	E: The same as in A except that $2\times$ Ca\textsuperscript{2+} is provided.
	F: The same as in E except with slowed Ca\textsuperscript{2+} diffusion and half amount of CaM.
	}
\label{fig10}
\end{figure}

To understand why frequency dependence changed, we examined the reaction occurrences over time on Layer1 edges (Fig~\ref{fig11}). We noticed that regardless of the input frequency, the reaction pathway choice deviates from the predominant pathway when Ca\textsuperscript{2+} levels become saturating. In the presence of moderate Ca\textsuperscript{2+} influx (Fig~\ref{fig11}A,B), CaM molecules follow the predominant pathway by choosing reaction paths according to affinities: bindings between N0C2 and CaMKII are more likely to occur than bindings of Ca\textsuperscript{2+} ions on N1 sites of CaM molecules. CaMKII binding with N0C2 dominates all types of CaMKII-CaM binding reactions (Fig~\ref{fig11}D,E,G). However, when Ca\textsuperscript{2+} influx becomes saturating, bindings of Ca\textsuperscript{2+} ions on N1 sites of CaM molecules dominate over the bindings between N0C2 and CaMKII (Fig~\ref{fig11}C). CaM molecules tend to stay in the Layer1 untill they get fully loaded by Ca\textsuperscript{2+}. As a result, CaMKII binding with N2C2 becomes dominant over all other CaM CaMKII binding reactions (Fig~\ref{fig11}F,H,I). This decision change gives rise to an altered pathway and eventually a suboptimal phosphorylation level. Given the same total amount of Ca\textsuperscript{2+}, 10 Hz influx saturates CaM in a shorter time than 5 Hz. Correspondingly, the switch of pathway choice is more dramatic for 10 Hz at $4.5\times$ Ca\textsuperscript{2+}, resulting in the reversed frequency perference as observed previously. 

\begin{figure}[!h]
	\caption{{\bf Change of pathway choice as Ca\textsuperscript{2+} influx increases}
	A: With $2\times$ Ca\textsuperscript{2+} input, reaction occurrences of CaM N0C2 binding Ca\textsuperscript{2+} at the N1 site and directly binding with CaMKII. Thick lines are for 10 Hz input, and thin lines are for 5 Hz input. B: The same as in A except that $3\times$ Ca\textsuperscript{2+} is given. C: The same as in C except that $4.5\times$ Ca\textsuperscript{2+} is given. D-F: Accumulated reaction occurrences for CaM CaMKII binding reactions under varying amount of total Ca\textsuperscript{2+} influx from $2\times$, $3\times$ to $4.5\times$. Ca\textsuperscript{2+} inputs are delivered at 5 Hz. G-I: The same as in D-F except that Ca\textsuperscript{2+} inputs are delivered at 10 Hz.
	}
\label{fig11}
\end{figure}

To validate our conclusion about the pathway decision change, we used a reduced reaction network to capture the observed frequency preference reversal (Fig~\ref{fig12}). The reduced network is derived from the initial steps in the whole Ca\textsuperscript{2+}-CaM-CaMKII network. We used the original Smoldyn to simulate the simple network, in order to demonstrate that the reversal of frequency preference is not an artifact of our modified simulator but is inherent to the network itself. We used 5 pulses of instantaneous Ca\textsuperscript{2+} release as the input and varied the amount of Ca\textsuperscript{2+} influx per pulse. The CaMKII molecules are modeled as monomers with an arbirary phosphorylation rate of \SI{1}{\per\s}. As expected, the reversal of frequency preference can be qualitatively captured by the simple reaction scheme (Fig~\ref{fig12}A). The 10 Hz stimulus becomes saturating and fails to generate more phosphorylation than 5 Hz when Ca\textsuperscript{2+} influx reaches 40000 ions per pulse.

\begin{figure}[!h]
	\caption{{\bf Change of pathway choice demonstracted using a reduced network}
	A: The reduced reaction network derived from the complete network. CaMKII are modeled as monomers and can undergo phosphorylation at an arbitrary rate \SI{1}{\per\s} as long as bound with a CaM N1C2. 
	B: Summary of phosphorylation levels in the presence of varying Ca\textsuperscript{2+} ions per pulse from 10000 ions/pulse to 50000 ions/pulse delivered at 5 Hz and 10 Hz.
	}
\label{fig12}
\end{figure}


\subsection*{The structural orgnization of CaMKII subunits}

Recent work on CaMKII holoenzyme structure suggests that how subunits are organized can affect the activation of the holoenzyme. In particular, whether subunits are arranged in a compact or an extended way can change the accessibility of Ca\textsuperscript{2+}-CaM to CaMKII. It is known that the structural arrangement is related to a linker region between a subunit's kinase domain and the holoenzyme central hub. Bayer et al.~\cite{Bayer:2002er} examined splice variants of $\beta$-CaMKII. These variants have identical kinase domains yet different linker lengths. They responded to Ca\textsuperscript{2+} oscillations differently, even though they showed no difference to prolonged Ca\textsuperscript{2+}-CaM input. In particular, for Ca\textsuperscript{2+} pulse input, the variants with longer-linker length exhibited a higher autophosphorylation rate. Another study by Chao et al.~\cite{Chao:2011iw} also indicated that the linker length affects the capability of a subunit's kinase domain to undock from the central hub. Undocking helps the subunit to release from autoinhibited state. Thus a longer linker can keep the kinase domain further away from the central hub, allowing a better chance for Ca\textsuperscript{2+}-CaM to access the binding site.

In our prototype model, holoenzyme subunits have distinct physical locations and are separated by \SI{8}{\nm} from its neighbors. To examine such a structural effect on CaMKII activation, we varied the radius of an one-ring shaped holoenzyme from \SI{5}{\nm} to \SI{15}{\nm}. The holoenzyme consists of 6 euqally spaced subunits. Neighbor subunits are spaced the same length as the holoenzyme radius. According to the literature~\cite{Chao:2011iw,Gaertner:2004jk}, a typical one-ring holoenzyme radius is within \SI{5}{\nm} to \SI{8}{\nm}. For each radius condition, we tested 15 simulation trials using the default Ca\textsuperscript{2+} influx delivered at 5 Hz. As shown in Fig~\ref{fig13}, there is a slight but distinguishable increase in phosphorylation level as the radius of the holoenzyme increases.

\begin{figure}[!h]
	\caption{{\bf Effects of holoenzyme size on CaMKII phosphorylation.}
	Phophorylation levels as a function of the holoenzyme radius presented in mean+standard deviation. Each radius condition is summarized from 15 trials.
	}
\label{fig13}
\end{figure}


\section*{Discussion}
In the study, we presented an efficient approach to simulate multi-subunit molecules and reaction network in detailed kinetics, as well as the effects of diffusion on the reaction network. We also introduced an intuitive approach to analyzing the pathway choices of a network based on the reaction history of a simulation, allowing us to grasp insights of a large network quickly.

Our findings agree with observations from previous studies. First, under physiological conditions when Ca\textsuperscript{2+} influx is moderate, CaM molecules partially loaded with Ca\textsuperscript{2+} are important for CaMKII activation. In particular, the pathway through N0C1, N0C2, and KN0C2 plays a crucial role. Second, factors such as CaM availablility~\cite{DeKoninck:1998wh} and Ca\textsuperscript{2+} diffusion can affect the CaMKII frequency response. Experiment studies suggest that free diffusible CaM molecules are higly limited \textit{in vivo}~\cite{2008BpJ....95.6002S,Persechini:2002tb,LubyPhelps:1995kl}. The limited CaM further implies a regulatory role of of many endogenous CaM binding proteins ~\cite{Skene:1990kf}. Third, intracellular crowding and spatial homogeneity can slow down molecule diffusion. For example, a recent biophysical study~\cite{2013PNAS..11015794T} suggests that the diffusion of Ca\textsuperscript{2+} ions can be reduced by ten times in a nanodomain around the Ca\textsuperscript{2+} channel mouth. This complication in Ca\textsuperscript{2+} diffusion may have substantial effects on CaMKII phosphorylation and frequency response. Fourth, we demonstrated that the holoenzyme size can affect the level of phosphorylation. This is consistent with the idea that the configuration of a holoenzyme, whether compact or extended, affects CaM molecules accessing the CaMKII subunits.

%chao etal, autoinhibited through dimerization;
%stratton etal, subunits exchange 
Our model does not contain Thr305/Thr306 phosphorylation and some newly discovered mechanisms of CaMKII, which may lead to more complicated activation pattern of CaMKII holoenzymes. For example, it has been found that there exists a compact autoinhibition state, which occurs through dimerization of adjacent subunits from top and bottom rings~\cite{Chao:2010fn}. Once Ca\textsuperscript{2+} bound to a dimerized subunit, the dimer dissembles and the two subunits are released. Another recent study indicated that phosphorylated CaMKII subunits can undergo subunits exchange to faciliate proprogating the activation triggered by Ca\textsuperscript{2+}-CaM~\cite{Stratton:2014ct}.

One technical challenge for particle-based simulation is to handle diffusion-limited reactions, especially in the presence of highly concetrated molecules. One recent experimental study~\cite{Faas:2011fna} estimated that the N sites of CaM are acting very fast to bind Ca\textsuperscript{2+}, much faster than previously cited for CaM-N lobe binding kinetics in modeling studies. These binding kinetics are in the diffusion-limited regime, rendering traditional mass action based methods inaccurate. However for these fast kinetics, there currently lacks an efficient method to simulate. If using the original Smoldyn algorithm, the simulation time step has to be considerablly reduced to obtain the correct steady state~\cite{Andrews:2015}. Another software package using an enhanced Green's Function Algorithm~\cite{vanZon:2005jd} can handle the high concentration diffusion-limited reactions accurately, but it takes an impractically long time to simulate. It is interesting to examine the Ca\textsuperscript{2+}-CaM-CaMKII using the fast kinetics parameters and we anticipate to address this question in future development.

\section*{Methods}
\subsection*{Simulator modifications}

We expanded the molecule data structure in Smoldyn to include complexes, molecules and binding sites. A complex may contain multiple molecules and a molecule may contain multiple binding sites. Reactions are specified between binding sites. Each binding site has binary states. For example, bound is coded as 1 and unbound as 0; phosphorylated as 1 and unphosphorylated as 0. Each molecule has a vector to store the states of binding sites. All reactions are stored in a hash table with reactants and their binding states as entry keys. A hash table is a data structure to store association arrays as well as rapid lookup. In our case, the reaction network can be considered as associations between reactants, therefore ideal to be impemented using a hash table.
CaM is an example of a molecule with multiple binding sites. The binding reactions involving the N and C lobes of a CaM can be coded as in Fig~\ref{fig1}B. A CaMKII holoenzyme is an example of a complex composed of two 6-subunits rings. Each subunit is a molecule containing binding sites for calmodulin and phosphorylation. Each ring has a radius of \SI{8}{\nm} and is separated from its direct neighbors at a fixed distance of \SI{8}{\nm} (estimated from~\cite{Gaertner:2004jk}). For simplicity, we modeled CaMKII holoenzymes as one ring of 6 subunits.

In the original version of Smoldyn, each reaction generates new molecules as products and reactant molecules are removed. In our case, since one molecule can have multiple binding sites and is potentially associated with multiple partners, entirely removing a molecule is not practical because other attached molecules would also be affected. In addition, removing and generating new molecules makes it difficult to track the reaction history of a molecule. Therefore, during reactions we do not remove molecules, but merely change molecule binding states and positions. Molecules bound together physically overlap and synchronize their locations automatically. When molecules diffuse, the diffusion coefficient is determined by the dominant molecule. For example, when Ca\textsuperscript{2+} and CaM are bound, the attached Ca\textsuperscript{2+} molecule diffuse with the CaM diffusion rate; similarly, CaM bound to a CaMKII subunit will diffuse with the CaMKII diffusion rate.

Macromolecules usually have multiple binding sites, and sometimes these sites compete for the same ligand. For example, CaM has 4 Ca\textsuperscript{2+} binding sites. Since the N and C sites act independently, the N1 and C1 sites compete for Ca\textsuperscript{2+} and an apoCaM N0C0 can become either N1C0 or N0C1, resulting in a branching reaction scheme. Thus a decision process is needed to choose a reaction path when a binding event occurs. 

To do this, consider the following two reactions
\begin{quote}
\begin{verbatim}
	rxn1: camN0C0 + ca <-> camN1C0
	rxn2: camN0C0 + ca <-> camN0C1
\end{verbatim}
\end{quote}

Rxn1 has a forward rate constant $k_{f1}$ in \SI{}{\per\micro\Molar\per\s} and a backward rate constant $k_{b1}$ \SI{}{\per\s}. Rxn2 has a similar rate constants $k_{f2}$ and $k_{b2}$. The two reactions can be viewed together as an equivalent rxn3, which has overall kinetics rates $k_{f3}$ and $k_{b3}$. According to the law of mass action, the reactions can be written as differential equations and we can obtain
\begin{equation}
	k_{f3}=k_{f1}+k_{f2}
\end{equation}

Smoldyn uses binding radii to implement second order reactions. If two molecules are spatially separated by a distance smaller than the corresponding binding radius, then the reaction proceeds. In Smoldyn, a special algorithm is used to calculate the binding radius, which depends on the kinetic rate constant, simulation time step and total diffusion rate of reactants. In case of a branched binding scheme sharing common reactants, we first calculate a binding radius $r_3$ based on $k_{f3}$. If the distance between a molecule pair is smaller than $r_3$, binding happens. To make a reaction choice, we generated a uniformly distributed random number from 0 to 1. If the number falls in the range $(0,\frac{kf_1}{kf_3}]$, then rxn1 is chosen; instead if the number falls in the range $(\frac{kf_1}{kf_3}, 1]$, we pick rxn2. Following this approach, the network can be kept consistent with the prediction by the mass action law. 

\subsection*{Reaction network}
We focus on the interactions among Ca\textsuperscript{2+}, CaM, and CaMKII. We first used a Ca\textsuperscript{2+}-CaM network for testing to confirm that modifications to the simulator were working properly. Then we added CaMKII holoenzymes to study the reaction network in details. We set up a cube-shaped model to represent a portion of a cell body (Fig~\ref{fig14}A). The cube has dimensions of \SI{1}{\um} in width and \SI{2}{\um} in depth. The top surface of the cube represents the cell membrane, reflective to all molecules. The four sides are also reflective. The bottom surface is partially absorbing to Ca\textsuperscript{2+} ions but reflective to CaM. This conservation of CaM molecules and CaMKII subunits guarantees a steady state initial condition. This partial absorption is a built-in feature in the original Smoldyn to resemble unbounded diffusion~\cite{Andrews:2009dr}.

\begin{figure}[!h]
	\caption{{\bf Model geometry and Ca\textsuperscript{2+} parameters.}
	A: Dimensions of the cube-shaped model. Ca\textsuperscript{2+} channels are located on the top surface. 
	B: $n_{inf}(v)$ and $\tau_n(v)$ used in the voltage-gated Ca\textsuperscript{2+} channels.
	C: The number channels N by fitting the single channel current to the GHK equation. N equals 16 by fitting. 
	}
\label{fig14}
\end{figure}

Voltage-gated Ca\textsuperscript{2+} channels (presumably L-type) are located on the top surface to provide Ca\textsuperscript{2+} influx. For simplicity, these channels are located at the center of the membrane. The channels open and close depending on a time-varying membrane voltage file generated from a neuron model (described below). CaMKII subunits (typically $\beta$-subunits) are uniformly present at a concentration of \SI{10}{\umol}. They are also immobilized so they are presumed as attached to actins~\cite{Li:2016cq}. Freely diffusible CaM molecules are uniformly distributed at a concentration of \SI{5}{\micro\Molar}. This is consistent with the notion that at the resting level, freely diffusible CaM molecules are considerably limited in number compared to their binding proteins~\cite{Tran:2003fs,2008BpJ....95.6002S,LubyPhelps:1995kl}. Table S2 lists all the reactions with corresponding kinetic parameters involved in the network. Kinetic parameters are integrated from various sources as noted in Table S2 and are adjusted to satisfy microscopic reversibility. 

\subsection*{Ca\textsuperscript{2+} input conditions}


We used a previously constructed multi-compartment neuron model to generate the somatic action potential trains at frequencies of 5 Hz and 10 Hz~\cite{Li:2014fv}. The action potentials are activated by synaptic stimulations, following a series of presynaptic release of neurotransmitters which was modeled with a probability profile measured in experiments~\cite{Grover:2009hb}. Presynaptic stimulations activate spine AMPA receptors, elevate the membrane potential and open NMDA receptors, allowing Ca\textsuperscript{2+} ions to enter the spines and depolarize the dendrite tree. The depolarization rapidly propagates toward the soma, initiating somatic action potentials. The pattern of action potentials follows the synaptic stimulation. We varied the intervals of synaptic release to generate action potentials in the form of theta-burst with different inter-bursts intervals, which provide different Ca\textsuperscript{2+} influx frequencies for the reaction network (Fig~\ref{fig15}A,C).

\begin{figure}[!h]
	\caption{{\bf Patterns of membrane voltage and Ca\textsuperscript{2+} influx. }
	A: A 5 Hz action potential trains voltage file generated from the NEURON model. 
	B: Ca\textsuperscript{2+} influx generated using the 5 Hz action potentials voltage file. Ca\textsuperscript{2+} influx is stochastic. There are spontaneous influx in the absence of action potentials. 
	C: 10 Hz action potential trains voltage file generated from the NEURON model. 
	D: Ca\textsuperscript{2+} influx generated using the 10 Hz action potential file. 
	E: Accumulated Ca\textsuperscript{2+} influx for 5Hz, 10 Hz, 8 and 16 channels respectively. The total Ca\textsuperscript{2+} amount is matched between Ca\textsuperscript{2+} influx files.
	}
\label{fig15}
\end{figure}

The neuron model was constructed using NEURON~\cite{Carnevale:2006iv} with a detailed morphology and ion channel conductances. Its membrane voltage output was applied to the Ca\textsuperscript{2+} reaction network model. Since the kinetics of L-type Ca\textsuperscript{2+} channels are relatively fast, the membrane potentials are unaffected by the channel activities. Thus the neuron model and the reaction network model can be safely decoupled. In the reaction network model, the stochastic Ca\textsuperscript{2+} channels open and close in response to the voltage input to provide Ca\textsuperscript influx. The voltage-depdendent opening and closing of these channels are modeled in a Hudgkin-Huxley way~\cite{Tuckwell:2012tt}. The rates of open and close are functions of membrane voltage and calculated using varialbles of $n$ and $\tau_n$, where $n$ describes voltage-dependent activation and $\tau_n$ is the time constant of $n$ (Fig~\ref{fig15}B). The channel obeys a simple first order kinetics as follows.

\begin{quote}
 \verb|CaL{gate==0} <-> CaL[gate=1]|.
\end{quote}
Then the following set of equations are used to decribe voltage-gated Ca\textsuperscript{2+} channels:
\begin{equation}rate_{open}(V)=\frac{n_{\inf}}{\tau_n}\end{equation}
\begin{equation}rate_{close}(V)=\frac{1-n_{\inf}}{\tau_n}\end{equation}
\begin{equation}n_{\inf}=\frac{1}{1+exp(\frac{V+V_{1/2}}{slope})}\end{equation}
\begin{equation}\tau_n=0.06+3\times\frac{\sqrt{0.45\times0.55}}{exp((V+V_{1/2})\times0.55/slope)+exp(-(V+V_{1/2})\times0.45/slope)}\end{equation}

In particular, $n_{\inf}$ is the steady state value of $n$. $V_{1/2}$ and $slope$ together describe the channel activation in response to voltage. In our model, $V_{1/2}$ equals \SI{-15}{\mV} and $slope$ equals 8.0.

The $rate_{open}$ and $rate_{close}$ are used to calculate conditional probabilities to determine the state of a channel for the next time step in the following way
\begin{equation}P(C|O)=1-exp(-rate_{close}(V)dt)\end{equation}
\begin{equation}P(O|C)=1-exp(-rate_{open}(V)dt)\end{equation}
\begin{equation}P(O|O)=1-P(C|O)\end{equation}

For each channel, at a given time, a probability is calculated based on the membrane voltage to decided whether a channel opens. If it opens, a varying number of Ca\textsuperscript{2+} ions are generated. To calculate how many ions, we used the following equation~\cite{2013PNAS..11015794T} to obtain the unitary current for a single channel
\begin{equation}i_{ca}=-g(V-V_s)\frac{\exp\frac{-(V-V_s)}{RT/zF}}{1-\exp\frac{-(V-V_s)}{RT/zF}}\end{equation}
where $g$ is chosen as \SI{5}{\pico\siemens} and $RT/zF$ equals \SI{12}{\milli\volt}. A current density is calculated using GHK current equation~\cite{citeulike:306134} assuming an extracellular [Ca\textsuperscript{2+}] of \SI{2}{\m\Molar} and an intracellular concentration of \SI{50}{\nano\Molar}. Since the membrane surface is \SI{1}{\um\square}, the current density is converted to a total current $I_{ghk}$ for the area. On the other hand, assuming a total number of N channels are present, for each channel $\frac{I_{ghk}}{N}$ a single channel current is $i_{ca}$. By fitting the total current $I_{ghk}$ and $N\times i_{ca}$, we obtained a N of 16 and a $V_s$ of \SI{-1.90955}{\milli\volt} (Fig~\ref{fig15}C). Using the $i_{ca}$, the number of ions entered during one time step is calcuated as $\frac{i_{ca}}{2e}dt$ (Fig~\ref{15}B,D), where 2 is the valence of Ca\textsuperscript{2+} and $e$ is the elementary charge. In order to guarantee a consistent amount of total Ca\textsuperscript{2+} for a given set of 5 Hz and 10 Hz voltage files, we generated 40 trials of Ca\textsuperscript{2+} influx files for each frequency and then selected the ones with equivalent total Ca\textsuperscript{2+} influx (Fig~\ref{fig15}E).


\section*{Conclusion}
\section*{Supporting information}

% For figure citations, please use "Fig" instead of "Figure".
% Place figure captions after the first paragraph in which they are cited.

% % Include only the SI item label in the paragraph heading. Use the \nameref{label} command to cite SI items in the text.
% \paragraph*{S1 Fig.}
% \label{S1_Fig}
% {\bf Bold the title sentence.} Add descriptive text after the title of the item (optional).

% \paragraph*{S1 File.}
% \label{S1_File}
% {\bf Lorem ipsum.}  Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 Appendix.}
% \label{S1_Appendix}
% {\bf Lorem ipsum.} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

\paragraph*{S1 Table.}
\label{S1_Table}
{\bf The total amount of Ca\textsuperscript{2+} influx of selected files for 5 Hz and 10 Hz.} 
\paragraph*{S2 Table.}
\label{S2_Table}
{\bf Kinetic parameters of all reactions.}
\paragraph*{S3 Table.}
\label{S3_Table}
{\bf Reaction characterizations.}

\section*{Acknowledgments}

\nolinenumbers

% Either type in your references using
% \begin{thebibliography}{}
% \bibitem{}
% Text
% \end{thebibliography}
%
% or
%
% Compile your BiBTeX database using our plos2015.bst
% style file and paste the contents of your .bbl file
% here. See http://journals.plos.org/plosone/s/latex for 
% step-by-step instructions.
% 
\bibliography{paper}{}

\end{document}
